### PR TITLE
New. ForceProtection. Added new module.

### PR DESCRIPTION
**Pull request by svfcode**
task https://doboard.com/1/task/9380

### CODE REVIEW

File: lib/Cleantalk/ApbctWP/Antispam/ForceProtection.php [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+     * @param string $content
+     * @return string
+     */
+    private function checkIframeForms($content)
+    {
+        $content = preg_replace_callback('#<iframe\b[^>]*>.*?<\/iframe>#is', function ($matches) {
+
+            $iframe = $matches[0];
+
+            if (preg_match('#<iframe\s+[^>]*src="([^"]*https:\/\/forms\.zohopublic\.com[^"]*)"[^>]*>#i', $iframe)) {
```
Comment (alexandergull) [inline lib/Cleantalk/ApbctWP/Antispam/ForceProtection.php:111]: I see this case is applicable only for zoho forms, what about other iframes?

File: js/src/apbct-public--1-force-protection.js [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    /**
+     * Decode forms
+     * @return {void}
+     */
+    decodeForms() {
+        let form;
+
+        this.wrappers.forEach((wrapper) => {
+            form = wrapper.querySelector('div.ct-encoded-form').dataset.encodedForm;
+            if (form) {
```
Comment (alexandergull) [inline js/src/apbct-public--1-force-protection.js:105]: Form should be of string type to make atob run, furthermore atob can throw a domexception if the string is invalid b64 encode. Maybe try/catch?

File: js/src/apbct-public--1-force-protection.js [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+     * @param {string} message
+     * @return {void}
+     */
+    showMessageForBot(message) {
+        let form;
+
+        this.wrappers.forEach((wrapper) => {
+            form = wrapper.querySelector('div.ct-encoded-form').dataset.encodedForm;
+            if (form) {
+                wrapper.outerHTML = '<div class="ct-encoded-form-forbidden">' + message + '</div>';
```
Comment (alexandergull) [inline js/src/apbct-public--1-force-protection.js:122]: Could we get message sanitized?

File: js/src/apbct-public-bundle.js [hunk 4]
```diff
@@ -1600,6 +1600,133 @@ function apbctCancelAutocomplete(element) {
     }));
 }
 
+if (ctPublic.data__key_is_ok) {
+    if (document.readyState !== 'loading') {
+        apbctForceProtect();
```
Comment (alexandergull) [inline js/src/apbct-public-bundle.js:1605]: Do we need run this independent of the surrent setting statement? Maybe localize 'forms__force_protection' to ctPublic? I can't see any ways to skip encode/decode on public JS.

File: js/src/apbct-public--1-force-protection.js [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+     */
+    checkBot() {
+        let data = {
+            event_javascript_data: getJavascriptClientData(),
+            post_url: document.location.href,
+            referrer: document.referrer,
+        };
+
+        if (ctPublicFunctions.data__ajax_type === 'rest') {
+            apbct_public_sendREST('force_protection_check_bot', {
```
Comment (alexandergull) [inline js/src/apbct-public--1-force-protection.js:55]: Why route name is different from ajax hook name? **apbct_force_protection_check_bot**  vs **force_protection_check_bot**.
Could we use the same naming?

**Review by alexandergull — CHANGES_REQUESTED** [review]

**Review by Glomberg — APPROVED** [review]

File: js/src/apbct-public--1-force-protection.js (hunk 2 above)
Comment (AntonV1211) [reply js/src/apbct-public--1-force-protection.js:105]: ok

File: js/src/apbct-public-bundle.js (hunk 4 above)
Comment (AntonV1211) [reply js/src/apbct-public-bundle.js:1605]: [https://github.com/CleanTalk/wordpress-antispam/pull/477/files#diff-322eb766244dd8d3a92d89fe40c02788fcc2d870af63f6cb4deb27526fa1bc57R193](url)
Here it is checked whether the option is enabled, if not, the wrapper will not load, therefore there will be no request

File: js/src/apbct-public--1-force-protection.js (hunk 5 above)
Comment (AntonV1211) [reply js/src/apbct-public--1-force-protection.js:55]: The prefix is added later

File: js/src/apbct-public--1-force-protection.js (hunk 3 above)
Comment (AntonV1211) [reply js/src/apbct-public--1-force-protection.js:127]: I think that sanitation is not needed here

**Review by alexandergull — APPROVED** [review]

**Pull request merged**